Concurrent FungibleAsset balance

igor-aptos opened the pull request:
### Description

### Test Plan
<!-- Please provide us with clear details for verifying that your changes work. -->

igor-aptos commented:
Current dependencies on/for this PR:
* main
  * **PR #11183** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11183?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11183?utm_source=stack-comment).

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -307,8 +328,12 @@ module aptos_framework::fungible_asset {
+    public fun balance<T: key>(store: Object<T>): u64 acquires FungibleStore, ConcurrentFungibleBalance {
+        let store_addr = object::object_address(&store);
+        if (concurrent_fungible_balance_exists(store_addr)) {
+            let balance = borrow_global<ConcurrentFungibleBalance>(store_addr);
```
I usually prefer
```
let balance = &borrow_global<ConcurrentFungibleBalance>(store_addr).balance;
aggregator_v2::read(balance)
```
because `balance.balance` looks super weird. Either things need to be called differently, or we can access the field directly when we borrow global

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -307,8 +328,12 @@ module aptos_framework::fungible_asset {
-        if (store_exists(object::object_address(&store))) {
+    public fun balance<T: key>(store: Object<T>): u64 acquires FungibleStore, ConcurrentFungibleBalance {
+        let store_addr = object::object_address(&store);
+        if (concurrent_fungible_balance_exists(store_addr)) {
```
We should check here that `&& store_exists(store_addr)` as well?

georgemitenkov replied:
This can be part of `concurrent_fungible_balance_exists` function?

igor-aptos replied:
that should be a general invariant, but not sure we want to/need to explicitly check it here?

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -366,16 +391,29 @@ module aptos_framework::fungible_asset {
+
+        if (default_to_concurrent_fungible_balance()) {
+            move_to(store_obj, ConcurrentFungibleBalance {
+                balance: aggregator_v2::create_unbounded_aggregator(),
```
Do we want unbounded here? Should `create_store` be aware of the limit?

igor-aptos replied:
it might be that limit can change. then it would be a nightmare if balances have different limits stored.

they are not minting/burning here, so checking against the limit is pointless. this method should never be able to overflow.

lightmark replied:
it's like u64... it has an implicit bound of U64::MAX, so does unbound aggregator.

georgemitenkov replied:
yes, but errors are different. For unbounded, it will be a nice move error, for u64::max it will be actually a runtime internal error on overflow. It is fine this way, but for consistency would be nice to have similar APIs and error types

davidiw replied:
code seems fine, "unbounded" only please, otherwise we get quirky behavior that doesn't exist in the basic FAS

also a better conversation for the AIP

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -366,16 +391,29 @@ module aptos_framework::fungible_asset {
             = move_from<FungibleStore>(addr);
         assert!(balance == 0, error::permission_denied(EBALANCE_IS_NOT_ZERO));
+
+        if (concurrent_fungible_balance_exists(addr)) {
```
I would prefer that we have `concurrent_fungible_balance_exists` before you do `assert!(balance == 0, error::permission_denied(EBALANCE_IS_NOT_ZERO));` after unpacking fungible store, otherwise we depend on the fact that the unused `balance` field is always zero? Something like `if .. else ..` reads better?

igor-aptos replied:
this is checking that both balances are 0, because we are only allowed to remove empty store, so current logic is correct?

georgemitenkov replied:
It is, this is why it is just a my preference (readability wise) but no strong opinion

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -386,22 +424,23 @@ module aptos_framework::fungible_asset {
             event::destroy_handle(withdraw_events);
             event::destroy_handle(frozen_events);
-        }
+
```
nit: remove this newline?

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -532,15 +571,21 @@ module aptos_framework::fungible_asset {
+            aggregator_v2::add(&mut balance.balance, amount);
+        } else {
+            let store = borrow_global_mut<FungibleStore>(store_addr);
+            store.balance = store.balance + amount;
```
We don't care about overflows? Interesting, why?

lightmark replied:
will later do sth like
https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/coin.move#L547
cc @junkil-park

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -532,15 +571,21 @@ module aptos_framework::fungible_asset {
-        let store = borrow_global_mut<FungibleStore>(store_addr);
-        store.balance = store.balance + amount;
+
+        if (concurrent_fungible_balance_exists(store_addr)) {
```
ditto about checking store exists and avoiding balance.balance

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -549,12 +594,19 @@ module aptos_framework::fungible_asset {
+            let balance = borrow_global_mut<ConcurrentFungibleBalance>(store_addr);
+            assert!(aggregator_v2::try_sub(&mut balance.balance, amount), error::invalid_argument(EINSUFFICIENT_BALANCE));
+        } else {
+            let store = borrow_global_mut<FungibleStore>(store_addr);
```
We don't check store exists here, is the check somewhere else?

lightmark replied:
yes. delegate to caller

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -656,6 +708,25 @@ module aptos_framework::fungible_asset {
+        let store_object_address = object::address_from_extend_ref(ref);
+        let store_object_signer = object::generate_signer_for_extending(ref);
+        assert!(features::concurrent_fungible_assets_enabled(), error::invalid_argument(ECONCURRENT_SUPPLY_NOT_ENABLED));
+        assert!(exists<FungibleStore>(store_object_address), error::not_found(ESTORE_NOT_FOUND));
```
reuse `store_exists`?

alnoki commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -656,6 +708,25 @@ module aptos_framework::fungible_asset {
     }
 
+    public fun upgrade_store_to_concurrent(
+        ref: &ExtendRef,
```
Unfortunately the requirement of an `ExtendRef` blocks upgrades in the general case, most importantly for wallet transfers.

I've solved this in https://github.com/aptos-labs/aptos-core/pull/12348

lightmark commented on aptos-move/framework/aptos-framework/doc/fungible_asset.md:
```diff
@@ -11,18 +11,32 @@ metadata object can be any object that equipped with <code><a href="fungible_ass
 -  [Resource `ConcurrentSupply`](#0x1_fungible_asset_ConcurrentSupply)
 -  [Resource `Metadata`](#0x1_fungible_asset_Metadata)
 -  [Resource `FungibleStore`](#0x1_fungible_asset_FungibleStore)
+<<<<<<< HEAD
```
rerun>?

alnoki replied:
Note this is fixed in #12348

alnoki commented on aptos-move/framework/aptos-framework/doc/fungible_asset.md:
```diff
@@ -2014,14 +2415,39 @@ Extract <code>amount</code> of the fungible asset from <code>store</code>.
-    <b>assert</b>!(store.balance &gt;= amount, <a href="../../aptos-stdlib/../move-stdlib/doc/error.md#0x1_error_invalid_argument">error::invalid_argument</a>(<a href="fungible_asset.md#0x1_fungible_asset_EINSUFFICIENT_BALANCE">EINSUFFICIENT_BALANCE</a>));
-    store.balance = store.balance - amount;
 
+<<<<<<< HEAD
```
Note this is fixed in #12348

igor-aptos commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -155,6 +162,14 @@ module aptos_framework::fungible_asset {
         frozen: bool,
     }
 
+    fun default_to_concurrent_fungible_supply(): bool {
```
in case we want to add multiple flags, to separately decide which path to take

alnoki commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -155,6 +163,18 @@ module aptos_framework::fungible_asset {
         frozen: bool,
     }
 
+    fun default_to_concurrent_fungible_supply(): bool {
+        features::concurrent_fungible_assets_enabled()
+    }
+
+    fun default_to_concurrent_fungible_balance(): bool {
+        features::concurrent_fungible_assets_enabled()
+    }
+
+    fun auto_upgrade_to_concurrent_fungible_balance(): bool {
+        features::concurrent_fungible_assets_enabled()
+    }
```
```suggestion

    inline fun default_to_concurrent_fungible_supply(): bool {
        features::concurrent_fungible_assets_enabled()
    }

    inline fun default_to_concurrent_fungible_balance(): bool {
        features::concurrent_fungible_assets_enabled()
    }

    inline fun auto_upgrade_to_concurrent_fungible_balance(): bool {
        features::concurrent_fungible_assets_enabled()
    }
```


How about making these all inline to reduce function calls on stack? 

The gas schedule charges for each function call, and these can easily be inlined

Since these are inner function for lots of functions across the framework stack, failing to optimizing them as much as possible here will have adverse consequences downstream

alnoki commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -289,6 +309,13 @@ module aptos_framework::fungible_asset {
         exists<FungibleStore>(store)
     }
 
+    #[view]
+    /// Return whether the provided address has a concurrent fungible balance initialized,
+    /// at the fungible store address.
+    public fun concurrent_fungible_balance_exists(store: address): bool {
+        exists<ConcurrentFungibleBalance>(store)
+    }
```
```suggestion
    #[view]
    /// Return whether the provided address has a concurrent fungible balance initialized,
    /// at the fungible store address.
    public fun concurrent_fungible_balance_exists(store: address): bool {
        exists<ConcurrentFungibleBalance>(store)
    }


    inline fun concurrent_fungible_balance_exists_inline(store: address): bool {
        exists<ConcurrentFungibleBalance>(store)
    }

    inline fun store_exists_inline(store: address): bool {
        exists<FungibleStore>(store)
    }
```

How about adding inline versions here as well, which can be used in control flow statements?

alnoki commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -307,8 +334,12 @@ module aptos_framework::fungible_asset {
-        if (store_exists(object::object_address(&store))) {
+    public fun balance<T: key>(store: Object<T>): u64 acquires FungibleStore, ConcurrentFungibleBalance {
+        let store_addr = object::object_address(&store);
+        if (concurrent_fungible_balance_exists(store_addr)) {
+            let balance_resource = borrow_global<ConcurrentFungibleBalance>(store_addr);
+            aggregator_v2::read(&balance_resource.balance)
+        } else if (store_exists(store_addr)) {
```
How about using `concurrent_fungible_balance_exists_inline` and `store_exists_inline` as suggested above to eliminate gas bloat?

lightmark replied:
you are so paranoid, hah

georgemitenkov replied:
@alnoki has a point, we have many single-block functions, which are extra costs and extra performance losses. Why not inline if we could

alnoki commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -366,16 +397,29 @@ module aptos_framework::fungible_asset {
         let FungibleStore { metadata: _, balance, frozen: _ }
             = move_from<FungibleStore>(addr);
         assert!(balance == 0, error::permission_denied(EBALANCE_IS_NOT_ZERO));
+
+        if (concurrent_fungible_balance_exists(addr)) {
+            let ConcurrentFungibleBalance { balance } = move_from<ConcurrentFungibleBalance>(addr);
```
```suggestion
    if (concurrent_fungible_balance_exists_inline(addr)) {
```

To reduce gas bloat?

alnoki commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -532,15 +576,22 @@ module aptos_framework::fungible_asset {
-        let store = borrow_global_mut<FungibleStore>(store_addr);
-        store.balance = store.balance + amount;
+
+        if (auto_upgrade_to_concurrent_fungible_balance() || concurrent_fungible_balance_exists(store_addr)) {
```
```suggestion
        if (auto_upgrade_to_concurrent_fungible_balance() || concurrent_fungible_balance_exists_inline(store_addr)) {
```

To reduce gas bloat

alnoki commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -549,12 +600,23 @@ module aptos_framework::fungible_asset {
-        assert!(store.balance >= amount, error::invalid_argument(EINSUFFICIENT_BALANCE));
-        store.balance = store.balance - amount;
 
+        if (auto_upgrade_to_concurrent_fungible_balance() || concurrent_fungible_balance_exists(store_addr)) {
```
```suggestion
        if (auto_upgrade_to_concurrent_fungible_balance() || concurrent_fungible_balance_exists_inline(store_addr)) {
```

To reduce gas bloat

alnoki reviewed (changes requested):
How about strategic inlining to reduce gas bloat?

A lot of these helper functions can easily be inlined, and failing to optimize them here will have adverse consequences downstream because they are inner calls for lots of call flows across the framework

(Each function call comes with a fixed cost, but inline functions are effectively free because they are essentially copy-pasted at compile time)

igor-aptos commented:
@alnoki I'll include inlines to individual functions, but I'll skip creating duplicate functions for that.

once @lightmark changes land as well, we will stress test this, and see if it makes any difference, seems probably minor.

igor-aptos commented:
I'll add more context on one design choice, and anyone has feedback/thoughts, let me know.

With the current code, once this flag is enabled, and stores get migrated, individual store will contain two resources in the resource group:
`FungibleStore` and `ConcurrentFungibleBalance`.

we cannot remove balance field from `FungibleStore` - so it will still be present there, just deprecated, and always 0.


Alternative would be to have create `ConcurrentFungibleStore`, to replace `FungibleStore`, with all the same fields, just a different type for the `balance`. That way, in the future `FungibleStore` is deprecated and doesn't exist onchain, and store only contains clean `ConcurrentFungibleStore` resource. 

That is in general doable, except for the `create_store` function, which currently returns `Object<FungibleStore>`. `Object` is just a wrapped address, but it asserts that object exists at that address, and so: 
- if someone has stored this reference in their code, doing migration internally would break it (I think?)
- we cannot change signature of `create_store`, so we would need to permanently support creating deprecated store, and then migrating it on next call, and having separate `create_concurrent_store` to be used instead.

alnoki commented:
> I'll add more context on one design choice, and anyone has feedback/thoughts, let me know.
> 
> With the current code, once this flag is enabled, and stores get migrated, individual store will contain two resources in the resource group: `FungibleStore` and `ConcurrentFungibleBalance`.
> 
> we cannot remove balance field from `FungibleStore` - so it will still be present there, just deprecated, and always 0.
> 
> Alternative would be to have create `ConcurrentFungibleStore`, to replace `FungibleStore`, with all the same fields, just a different type for the `balance`. That way, in the future `FungibleStore` is deprecated and doesn't exist onchain, and store only contains clean `ConcurrentFungibleStore` resource.
> 
> That is in general doable, except for the `create_store` function, which currently returns `Object<FungibleStore>`. `Object` is just a wrapped address, but it asserts that object exists at that address, and so:
> 
> * if someone has stored this reference in their code, doing migration internally would break it (I think?)
> * we cannot change signature of `create_store`, so we would need to permanently support creating deprecated store, and then migrating it on next call, and having separate `create_concurrent_store` to be used instead.
> 
> Any thoughts?

@igor-aptos 

1. One option is to just change the internals of `create_store` to `abort 0`: this is backwards compatible. Then you could have `create_concurrent_store` which uses the `ConcurrentFungibleStore` you mentioned
2. For the on-the-fly upgrade you could just `move_from` the `FungibleStore`, destruct it, then `move_to` the `ConcurrentFungibleStore`
3. Operations will need to `ensure_store_migrated_to_concurrent` as applicable
4. Users who have a reference to `FungibleStore` in their code will indeed have breaking changes. Simple example, someone just checks if the object exists, and if not, tries to create one. Before upgrade the check will say that ther eis no store, but after the upgrade the logic will say there is no store so they'll probably try to create a new one, which will error out because then the `ConcurrentFungibleStore` would've already been created

lightmark commented:
> > I'll add more context on one design choice, and anyone has feedback/thoughts, let me know.
> > With the current code, once this flag is enabled, and stores get migrated, individual store will contain two resources in the resource group: `FungibleStore` and `ConcurrentFungibleBalance`.
> > we cannot remove balance field from `FungibleStore` - so it will still be present there, just deprecated, and always 0.
> > Alternative would be to have create `ConcurrentFungibleStore`, to replace `FungibleStore`, with all the same fields, just a different type for the `balance`. That way, in the future `FungibleStore` is deprecated and doesn't exist onchain, and store only contains clean `ConcurrentFungibleStore` resource.
> > That is in general doable, except for the `create_store` function, which currently returns `Object<FungibleStore>`. `Object` is just a wrapped address, but it asserts that object exists at that address, and so:
> > 
> > * if someone has stored this reference in their code, doing migration internally would break it (I think?)
> > * we cannot change signature of `create_store`, so we would need to permanently support creating deprecated store, and then migrating it on next call, and having separate `create_concurrent_store` to be used instead.
> > 
> > Any thoughts?
> 
> @igor-aptos
> 
> 1. One option is to just change the internals of `create_store` to `abort 0`: this is backwards compatible. Then you could have `create_concurrent_store` which uses the `ConcurrentFungibleStore` you mentioned
> 2. For the on-the-fly upgrade you could just `move_from` the `FungibleStore`, destruct it, then `move_to` the `ConcurrentFungibleStore`
> 3. Operations will need to `ensure_store_migrated_to_concurrent` as applicable
> 4. Users who have a reference to `FungibleStore` in their code will indeed have breaking changes. Simple example, someone just checks if the object exists, and if not, tries to create one. Before upgrade the check will say that ther eis no store, but after the upgrade the logic will say there is no store so they'll probably try to create a new one, which will error out because then the `ConcurrentFungibleStore` would've already been created

1. You assume nobody has `create_store` in their code. Otherwise, it breaks a lot.
2. This may need `create_signer`, but sounds acceptable.
3. n/a
4. This is not a main issue. The main issue is if someone call `object::address_to_object<FungibleStore>` in their code, it will abort.
Due to 1 and 4, I vote for the current solution.

lightmark commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -656,6 +718,20 @@ module aptos_framework::fungible_asset {
+    ) {
+        if (!exists<ConcurrentFungibleBalance>(fungible_store_address)) {
+            let store = borrow_global_mut<FungibleStore>(fungible_store_address);
+            let balance = aggregator_v2::create_unbounded_aggregator();
```
maybe we should support initialize with a value.

igor-aptos replied:
yeah, if we started with that, we could've had one native and one utility functions. Seems like we will have to have two native functions now :/

lightmark commented:
btw, this is not final right? we have to benchmark to decide whether we want people to opt-in.

igor-aptos commented:
@lightmark it's gated  - so we can land it,  and then benchmarl, and update only flag if needed?

davidiw commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -549,12 +609,26 @@ module aptos_framework::fungible_asset {
+            ensure_store_upgraded_to_concurrent_internal(store_addr);
+        };
+
+        if (concurrent_fungible_balance_exists(store_addr)) {
```
why are we happy pathing aggregator

igor-aptos replied:
with making it opt-in , updated to happy path non-aggregator

davidiw commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -656,6 +730,30 @@ module aptos_framework::fungible_asset {
+    inline fun ensure_store_upgraded_to_concurrent_internal(
+        fungible_store_address: address,
+    ) acquires FungibleStore {
+        if (!exists<ConcurrentFungibleBalance>(fungible_store_address)) {
```
why the nest... can't we just exit or assert this hasn't been called... so said to see this type of nesting... I have to read the end of a massive blob to see there's no else.

alnoki replied:
@davidiw thanks for flagging. The below suggestion eliminates a bytecode instruction:

```rust
    /// Ensure a known `FungibleStore` has `ConcurrentFungibleBalance`.
    inline fun ensure_store_upgraded_to_concurrent_internal(
        fungible_store_address: address,
    ) acquires FungibleStore {
        if (exists<ConcurrentFungibleBalance>(fungible_store_address)) {
            // Do nothing, return statements not currently supported for inline.
        } else {
            let store = borrow_global_mut<FungibleStore>(fungible_store_address);
            let balance = aggregator_v2::create_unbounded_aggregator();
            aggregator_v2::add(&mut balance, store.balance);
            store.balance = 0;
            let object_signer = create_signer::create_signer(fungible_store_address);
            move_to(&object_signer, ConcurrentFungibleBalance { balance });
        };
    }
```

# Current bytecode result from1c38359fa9

Taken from the disassembly of `upgrade_store_to_concurrent()`

```asm
	27: Exists[1](ConcurrentFungibleBalance)
	28: Not ; This is the instruction that gets eliminated
	29: BrFalse(51)
B10: ; This is where `ensure_store_upgraded_to_concurrent_internal()` begins
	30: CopyLoc[3](loc1: address)
	31: MutBorrowGlobal[10](FungibleStore)
	32: StLoc[5](loc3: &mut FungibleStore)
```

# Improved bytecode result from suggestion

Yields

```asm
	27: Exists[1](ConcurrentFungibleBalance)
	28: BrFalse(30)
B10:
	29: Branch(51)
B11: ; This is where `ensure_store_upgraded_to_concurrent_internal()` begins
	30: CopyLoc[3](loc1: address)
	31: MutBorrowGlobal[10](FungibleStore)
	32: StLoc[5](loc3: &mut FungibleStore)
```

# Steps to reproduce

```sh

```

igor-aptos replied:
ensure_store_upgraded_to_concurrent_internal is not going to be called often, with making it opt-in, so leaving it readable

davidiw reviewed (commented):
overall seems fine, let's follow up tomorrow on what the plan is?

igor-aptos commented:
Updated the PR for review comments and made assumption that feature will be opt-in, making non-concurrent path be faster (i.e. not needing to fetch ConcurrentFungibleBalance)

From perf impact:
* if we have it as opt-in, we have minuscule impact be 1-2% on non-conflicting workload, and huge perf improvement from parallelization if there are conflicts (i.e. single fee payer account)
* if we make it default for everyone, impact looks like 4-6% regression on non-conflicting workload. (to avoid slightly worse regression I [duplicate the metadata](https://github.com/aptos-labs/aptos-core/pull/13189/files#diff-7026c328622536165825a6f5dcaf13764bd475489ecba30ba4716dfdc4acc1e9) to not require fetching FungibleStore resource at all, unfortunately, I cannot replace FungibleStore with ConcurrentFungibleStore, as people/API use Object<FungibleStore> which checks for existence of FungibleStore)

@movekevin was mentioning against inlining non-trivial functions, so removed inline on autoconversion.

alnoki commented:
> @movekevin was mentioning against inlining

For the current situation in which there are no compiler optimizations, eliminating inlining will only inhibit performance. The inlined code is easy enough to test in all paths, and there is no issue of making the bytecode larger because it is effectively free to increase the package size of the Aptos Framework as a first-class library. 

The current issues with the feature are all about performance, and inlining all functions where possible is a guaranteed way to solve performance

@movekevin why are you advocating against a performance increase?

igor-aptos commented:
There are three options being discussed for upgrading balance field. Current struct is:

```
    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
    struct FungibleStore has key {
        metadata: Object<Metadata>,
        balance: u64,
        frozen: bool,
    }
```


Option 1, move balance to separate resource:
```
    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
    struct ConcurrentFungibleBalance has key {
        balance: Aggregator<u64>,
    }
```
    
Option 2, more cumbersome, more performant, copy other fields as well:

```
    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
    struct ConcurrentFungibleStore has key {
        metadata: Object<Metadata>,
        balance: Aggregator<u64>,
        frozen: bool,
    }
    
```
Option 3, allow u64 => Aggregator*<u64> upgradeability:

```
    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
    struct FungibleStore has key {
        metadata: Object<Metadata>,
        balance: Aggregator*<u64>,
        frozen: bool,
    }
```

alnoki commented:
> Option 3, allow u64 => Aggregator* upgradeability:

@igor-aptos If this could be done in a backwards-compatible manner then it would probably streamline things the most, however it would generate some breaking API changes at least with the current Move VM because it modifies an existing struct, correct?

Assuming that option 3 requires heavy modifications to the upgradeability paradigm, the option 2 approach should help to eliminate some of the call path inefficiencies related to borrows if implemented as I suggested here: https://github.com/aptos-labs/aptos-core/pull/13203#issuecomment-2094448647

> Option 2, more cumbersome, more performant, copy other fields as well:
> 
> ```
>     #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
>     struct ConcurrentFungibleStore has key {
>         metadata: Object<Metadata>,
>         balance: Aggregator<u64>,
>         frozen: bool,
>     }
>     
> ```

igor-aptos commented:
Updated PR, ready for review:

* we will be going with opt-in, so enabling CONCURRENT_FUNGIBLE_BALANCE. Other flag - DEFAULT_TO_CONCURRENT_FUNGIBLE_BALANCE will not be enabled, but is provided and implemented for future testing purposes.
* so happy path is non-concurrent one, based on balance != 0 check
* in the future, if this is useful, we can consider changing balance field in FungibleAsset be Aggregator, which will convert everyone.

georgemitenkov commented on aptos-move/framework/aptos-framework/sources/fungible_asset.move:
```diff
@@ -82,6 +83,8 @@ module aptos_framework::fungible_asset {
     const EFUNGIBLE_METADATA_EXISTENCE: u64 = 30;
     /// Cannot register dispatch hook for APT.
     const EAPT_NOT_DISPATCHABLE: u64 = 31;
+    /// Flag for Concurrent Supply not enabled
```
nit: . is missing and it should be balance, not supply?

alnoki replied:
```suggestion
Flag for Concurrent balance not enabled.
```
